torchvision/datasets/video_utils.py

```diff
@@ -0,0 +1,188 @@
+    """
+    assert tensor.dim() == 1
+    o_stride = tensor.stride(0)
+    numel = tensor.numel()
+    new_stride = (step * o_stride, dilation * o_stride)
+    new_size = ((numel - (dilation * (size - 1) + 1)) // step + 1, size)
+    if new_size[0] < 1:
+        new_size = (0, size)
+    return torch.as_strided(tensor, new_size, new_stride)
+
+
+class VideoClips(object):
+    """
+    Given a list of video files, computes all consecutive subvideos of size
+    `clip_length_in_frames`, where the distance between each subvideo in the
+    same video is defined by `frames_between_clips`.
+    If `frame_rate` is specified, it will also resample all the videos to have
+    the same frame rate, and the clips will refer to this frame rate.
+
+    Creating this instance the first time is time-consuming, as it needs to
+    decode all the videos in `video_paths`. It is recommended that you
+    cache the results after instantiation of the class.
+
+    Recreating the clips for different clip lengths is fast, and can be done
+    with the `compute_clips` method.
+
+    Arguments:
+        video_paths (List[str]): paths to the video files
+        clip_length_in_frames (int): size of a clip in number of frames
+        frames_between_clips (int): step (in frames) between each clip
+        frame_rate (int, optional): if specified, it will resample the video
+            so that it has `frame_rate`, and then the clips will be defined
+            on the resampled video
+    """
+    def __init__(self, video_paths, clip_length_in_frames=16, frames_between_clips=1,
+                 frame_rate=None):
+        self.video_paths = video_paths
+        self._compute_frame_pts()
+        self.compute_clips(clip_length_in_frames, frames_between_clips, frame_rate)
+
+    def _compute_frame_pts(self):
+        self.video_pts = []
+        self.video_fps = []
+        # TODO maybe paralellize this
+        for video_file in self.video_paths:
+            clips, fps = read_video_timestamps(video_file)
+            self.video_pts.append(torch.as_tensor(clips))
+            self.video_fps.append(fps)
+
+    def compute_clips(self, num_frames, step, frame_rate=None):
+        """
+        Compute all consecutive sequences of clips from video_pts.
+
+        Arguments:
+            num_frames (int): number of frames for the clip
+            step (int): distance between two clips
+            dilation (int): distance between two consecutive frames
+                in a clip
+        """
+        self.num_frames = num_frames
+        self.step = step
+        self.frame_rate = frame_rate
+        self.clips = []
+        for video_pts, fps in zip(self.video_pts, self.video_fps):
+            if frame_rate is not None:
+                # divup, == int(ceil(fps / frame_rate))
+                dilation = max(int((fps + frame_rate - 1) // frame_rate), 1)
+            else:
+                dilation = 1
+            clips = unfold(video_pts, num_frames, step, dilation)
```

Comment:
unfold will drop the last frames of the video, if the last clips is smaller than the other.
It would be nice to have an option to pad the last clip.

Reply:
This is a great point. We might add a `drop_last` option to `VideoClips`, as what we have in `BatchSampler`. But it will be slightly more annoying to implement, because we won't be able to use stride tricks anymore.

Reply:
Created an issue to track this request in https://github.com/pytorch/vision/issues/1113

Reply:
For now, I've mentioned in the documentation that the last frames of a video can potentially be dropped.

I'm going to be merging this as is for now, but I'll try to address this before the release.